src/client/pythonEnvironments/collection/environmentsReducer.ts

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { cloneDeep, isEqual } from 'lodash';
+import { Event, EventEmitter } from 'vscode';
+import { traceVerbose } from '../../common/logger';
+import { createDeferred } from '../../common/utils/async';
+import { areSameEnvironment, PythonEnvInfo, PythonEnvKind } from '../base/info';
+import {
+    ILocator, IPythonEnvsIterator, PythonEnvUpdatedEvent, QueryForEvent,
+} from '../base/locator';
+import { PythonEnvsChangedEvent } from '../base/watcher';
+
+/**
+ * Combines duplicate environments received from the incoming locator into one and passes on unique environments
+ */
+export class PythonEnvsReducer implements ILocator {
+    public get onChanged(): Event<PythonEnvsChangedEvent> {
+        return this.pythonEnvsManager.onChanged;
+    }
+
+    constructor(private readonly pythonEnvsManager: ILocator) {}
+
+    public async resolveEnv(env: string | PythonEnvInfo): Promise<PythonEnvInfo | undefined> {
+        let environment: PythonEnvInfo | undefined;
+        const waitForUpdatesDeferred = createDeferred<void>();
+        const iterator = this.iterEnvs();
+        iterator.onUpdated!((event) => {
+            if (event === null) {
+                waitForUpdatesDeferred.resolve();
+            } else if (environment && areSameEnvironment(environment, event.new)) {
+                environment = event.new;
+            }
+        });
+        let result = await iterator.next();
+        while (!result.done) {
+            if (areSameEnvironment(result.value, env)) {
+                environment = result.value;
+            }
+            // eslint-disable-next-line no-await-in-loop
+            result = await iterator.next();
+        }
+        if (!environment) {
+            return undefined;
+        }
+        await waitForUpdatesDeferred.promise;
+        return this.pythonEnvsManager.resolveEnv(environment);

```

Comment:
As noted when we chatted, this is much simpler if you use `getEnvs()` (from #14051):

```suggestion
        const iterator = this.iterEnvs();
        const envs = await getEnvs(iterator);
        const matched = typeof env === 'string'
            ? envs.find((e) => areSameFile(env, e.executable.filename))
            : envs.find((e) => areSameEnvironment(env, e));
        return this.pythonEnvsManager.resolveEnv(matched);
```

As we also discussed, iterating over the whole list isn't very efficiient but probably unavodiable.

Reply:
Yup, I'll include this and remove all the dummy implementations later in one go, will create a task for that. Btw we can merge the needs for `areSameFile` within `areSameEnvironment` as well, I asked Karthik to do that yesterday.